File: Problems & Solutions/1833. Maximum Ice Cream Bars/06 January 2022/Approach 2 - Counting Sort (Greedy)/Maximum Ice Cream Bars.java
class Solution {
    public int maxIceCream(int[] costs, int coins) {
        int n = costs.length;
        int icecreams = 0;

        int m = costs[0];
        for (int cost : costs) {
            m = Math.max(m, cost);
        }

        int[] costsFrequency = new int[m + 1];
        for (int cost : costs) {
            costsFrequency[cost]++;
        }

        for (int cost = 1; cost <= m; ++cost) {
            if (costsFrequency[cost] == 0) {
                continue;
            }
            if (coins < cost) {
                break;
            }
            
            int count = Math.min(costsFrequency[cost], coins / cost);
            coins -= cost * count;
            icecreams += count;
        }

        return icecreams;
    }
}

File: Problems & Solutions/1833. Maximum Ice Cream Bars/06 January 2022/Approach 2 - Counting Sort (Greedy)/Maximum Ice Cream Bars.js
var maxIceCream = function(costs, coins) {
    let n = costs.length;
    let icecreams = 0;
    let m = Math.max(...costs);

    let costsFrequency = Array(m + 1).fill(0);
    for (let cost of costs) {
        costsFrequency[cost] += 1;
    }

    for (let cost = 1; cost <= m; ++cost) {
        if (!costsFrequency[cost]) {
            continue;
        }
        if (coins < cost) {
            break;
        }
        
        let count = Math.min(costsFrequency[cost], Math.floor(coins / cost));
        coins -= cost * count;
        icecreams += count;
    }

    return icecreams;
};